src/Events.Functions/EventsRegistration.cs

```diff
@@ -0,0 +1,52 @@
+using System.Text.Json;
+using System.Threading.Tasks;
+using Altinn.Platform.Events.Functions.Clients.Interfaces;
+using Altinn.Platform.Events.Functions.Models;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Extensions.Logging;
+
+namespace Altinn.Platform.Events.Functions
+{
+    /// <summary>
+    /// Process incoming CloudEvents in "events-registration" queue.
+    /// CloudEvents are first saved to the database
+    /// before being added to the "events-inbound" queue.
+    /// </summary>
+    public class EventsRegistration
+    {
+        private readonly IEventsStorageClient _eventsStorageClient;
+        private readonly IInboundClient _inboundQueueClient;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventsInbound"/> class.
+        /// </summary>
+        public EventsRegistration(
+            IEventsStorageClient eventsStorageClient,
+            IInboundClient inboundQueueClient)
+        {
+            _eventsStorageClient = eventsStorageClient;
+            _inboundQueueClient = inboundQueueClient;
+        }
+
+        /// <summary>
+        /// Saves cloudEvents from events-registration queue to persistent storage
+        /// and sends to events-inbound queue storage.
+        /// </summary>
+        [FunctionName("EventsRegistration")]
+#pragma warning disable IDE0060 // Remove unused parameter
+        public async Task Run([QueueTrigger("events-registration", Connection = "QueueStorage")] string item, ILogger log)
+#pragma warning restore IDE0060 // Remove unused parameter
+        {
+            CloudEvent cloudEvent = JsonSerializer.Deserialize<CloudEvent>(item);
+
+            /*
+            Attempt to save cloudEvent. 
+            If saving fails, the cloudEvent will automatically be 
+            returned to events-registration queue for retry handling.
+            */
+
+            await _eventsStorageClient.SaveCloudEvent(cloudEvent);
```

Comment:
Do we need to handle the potential throw from SaveCloudEvent? eventsStorageClient will log something, but do we need to handle this in the function as well?

Reply:
I am concerned we already have logging in triplicate here already. In the case of an http error in the target api, our handlers will log two (identical) errors and the QueueTrigger runtime will also log something. 
On second look, I think we could simplify all clients if we isolate logging of AccessToken generation errors in the base class.

Reply:
I don't think we need any extra logging or handling here as the event will end up back on the queue.. I do however question the extra load on the events component if the queue is the error source. The component might be idempotent, but do we want this extra load on the solution if the event has already been persisted successfully ?

Reply:
I think we concluded that if the target queue is unavailable, then the source queue will also be unavailable and our function will not be triggered. Can you think of any scenario where that wouldn't be the case?